File: mobile application development/front/kotlin/app/src/main/java/com/itmo/mybroker/core/PortfolioStore.kt
package com.itmo.mybroker.core

import com.itmo.mybroker.api.TradeResponse
import com.itmo.mybroker.data.BrokerData
import com.itmo.mybroker.data.Ccy

data class ExecutedTrade(
    val id: String,
    val instrumentId: String,
    val symbol: String,
    val side: String,
    val quantity: Double,
    val price: Double,
    val currency: Ccy,
    val executedAt: Long,
)

object PortfolioStore {
    private val trades = mutableListOf<ExecutedTrade>()

    fun seedIfEmpty() {
        if (trades.isNotEmpty()) return
        BrokerData.repository.portfolioHoldings.forEachIndexed { i, h ->
            val tk = BrokerData.repository.bySym(h.sym)
            trades.add(
                ExecutedTrade(
                    id = "seed-${h.sym}",
                    instrumentId = h.sym,
                    symbol = h.sym,
                    side = "buy",
                    quantity = h.qty.toDouble(),
                    price = h.avg,
                    currency = tk.ccy,
                    executedAt = java.util.Calendar.getInstance().apply { set(2026, 4, 1) }.timeInMillis + i * 86400000L,
                ),
            )
        }
    }

    fun recordTrade(t: TradeResponse) {
        val ccy = runCatching { Ccy.valueOf(t.currency) }.getOrDefault(Ccy.RUB)
        trades.add(
            ExecutedTrade(
                id = t.id,
                instrumentId = t.instrumentId,
                symbol = t.symbol,
                side = t.side,
                quantity = t.quantity,
                price = t.price,
                currency = ccy,
                executedAt = t.executedAt,
            ),
        )
    }

    fun getQuantity(sym: String): Double {
        val s = sym.uppercase()
        return trades.filter { it.symbol.uppercase() == s }
            .sumOf { if (it.side == "buy") it.quantity else -it.quantity }
    }

    fun getCostBasis(sym: String): Double? {
        val s = sym.uppercase()
        var qty = 0.0
        var cost = 0.0
        val sorted = trades.filter { it.symbol.uppercase() == s }.sortedBy { it.executedAt }
        for (t in sorted) {
            if (t.side == "buy") {
                cost += t.price * t.quantity
                qty += t.quantity
            } else {
                if (qty <= 0) continue
                val avg = cost / qty
                val sellQty = minOf(t.quantity, qty)
                cost -= avg * sellQty
                qty -= sellQty
            }
        }
        if (qty <= 0) return null
        return cost / qty
    }
}

File: mobile application development/front/react-native/src/stores/portfolioStore.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { create } from 'zustand';
import { createJSONStorage, persist } from 'zustand/middleware';
import type { Ccy } from '../data/market';
import type { TradeResponse } from '../api/types';
import { brokerData } from '../data/repository';

export type ExecutedTrade = {
  id: string;
  instrumentId: string;
  symbol: string;
  side: 'buy' | 'sell';
  quantity: number;
  price: number;
  currency: Ccy;
  executedAt: number;
};

type PortfolioState = {
  trades: ExecutedTrade[];
  recordTrade: (trade: TradeResponse) => void;
  seedFromMockIfEmpty: () => void;
  getCostBasis: (symbol: string) => number | null;
  getQuantity: (symbol: string) => number;
};

function toExecuted(t: TradeResponse): ExecutedTrade {
  return {
    id: t.id,
    instrumentId: t.instrumentId,
    symbol: t.symbol,
    side: t.side,
    quantity: t.quantity,
    price: t.price,
    currency: t.currency,
    executedAt: t.executedAt,
  };
}

function computeCostBasis(trades: ExecutedTrade[], symbol: string): number | null {
  const sym = symbol.toUpperCase();
  let qty = 0;
  let cost = 0;
  const sorted = [...trades]
    .filter((t) => t.symbol.toUpperCase() === sym)
    .sort((a, b) => a.executedAt - b.executedAt);

  for (const t of sorted) {
    if (t.side === 'buy') {
      cost += t.price * t.quantity;
      qty += t.quantity;
    } else {
      if (qty <= 0) continue;
      const avg = cost / qty;
      const sellQty = Math.min(t.quantity, qty);
      cost -= avg * sellQty;
      qty -= sellQty;
    }
  }

  if (qty <= 0) return null;
  return cost / qty;
}

function computeQuantity(trades: ExecutedTrade[], symbol: string): number {
  const sym = symbol.toUpperCase();
  return trades
    .filter((t) => t.symbol.toUpperCase() === sym)
    .reduce((q, t) => q + (t.side === 'buy' ? t.quantity : -t.quantity), 0);
}

export const usePortfolioStore = create<PortfolioState>()(
  persist(
    (set, get) => ({
      trades: [],

      recordTrade: (trade) => {
        set((s) => ({ trades: [...s.trades, toExecuted(trade)] }));
      },

      seedFromMockIfEmpty: () => {
        if (get().trades.length > 0) return;
        const seeded: ExecutedTrade[] = brokerData.portfolioHoldings.map((h, i) => {
          const tk = brokerData.findTicker(h.sym);
          return {
            id: `seed-${h.sym}`,
            instrumentId: h.sym,
            symbol: h.sym,
            side: 'buy' as const,
            quantity: h.qty,
            price: h.avg,
            currency: tk.ccy,
            executedAt: Date.UTC(2026, 4, 1) + i * 86400000,
          };
        });
        set({ trades: seeded });
      },

      getCostBasis: (symbol) => computeCostBasis(get().trades, symbol),

      getQuantity: (symbol) => computeQuantity(get().trades, symbol),
    }),
    {
      name: 'portfolio-trades',
      storage: createJSONStorage(() => AsyncStorage),
      partialize: (s) => ({ trades: s.trades }),
    }
  )
);
